### PR TITLE
(WIP) Custom model loading hooks

### PR DESCRIPTION
**Pull request by asiekierka**
Do not merge yet. This is for initial review - please also raise additional needs, if necessary.

### CODE REVIEW

**Comment by therealfarfetchd** [comment]
Ah, I get what you mean. So you want to be able to load a different model from the CustomModelLoader? Yeah, that might be a thing that needs to be considered.

Maybe pass a Function<ModelIdentifier, UnbakedModel> to the load method, that when called, tries to load a different model (of course, stuff like recursion detection would need to be implemented, that it doesn't call itself with the same ID twice per "top level" load)

**Comment by asiekierka** [comment]
Mojang already implements recursion detection, so that's a non-issue. I'm just thinking if that's the best way to do it - perhaps I need to sleep on it.

**Comment by Daomephsta** [comment]
One thing I'd request is decent support for extending other model format. The current system would allow that, but I'd have to add an affix to the identifier for the extended version, otherwise it would recurse infinitely or until recursion safeguards kicked in. 

That's a bit messy, and I think the situation could be solved fairly simply by adding a way to ask for a model to be loaded, ignoring the "active" loader. Perhaps as a default method on the interface. Some kind of collection of ignored loaders for the current model being loaded would have to exist though, as loaders farther down the loading chain might want to use the same feature. 

I can make a mockup of this if that would explain it better.

**Comment by asiekierka** [comment]
@Daomephsta I see what you mean; but I think that's a bit of a hack, really! The better approach would likely be to simply extend another mod's model as a separate file...

**Comment by Daomephsta** [comment]
@asiekierka 
How exactly would that work? The custom model loaders don't, and shouldn't, have any concept of inheritance.

**Comment by asiekierka** [comment]
@Daomephsta Except vanilla's model loader already *has* a model of inheritance. This is nothing new.

**Comment by asiekierka** [comment]
One of Player's proposals is replacing accepts+load with a @Nullable load instead, as "it's a potential source of problems" otherwise + presumably as it could save a little bit of performance at scale (think in a "not having a duplicate contains/get" kind of way, except with resource resolution).

Another proposal I received is only checking trailing loaders optionally/in development, but the problem here is that issues caused by duplicate loaders are often non-obvious (rendering glitches, etc.) and hard to narrow down just by looking at a mod list, so I'm not keen on that.

Player also argues ModelRequester requires further thought, so if you have any ideas let me know!

**Comment by grondag** [comment]
This looks OK for model loading.  Unfortunately there is no way to handle truly dynamic, run-time model baking in 1.14 and even if Fabric were going to provide that, this is not where it belongs.

Mojang still uses an IdentifyHashMap with BlockState values as keys (the map lives in BlockModels) and at end of baking, every possible block state that may be rendered (without a BlockEntityRenderer) needs to be included in that hash map.  And it appears there is no longer any concept of "ActualState" distinct from the BlockState used for the model lookup. 

This is a completely rational design decision by Mojang - it ensures all model loading and baking occurs before the main render loop starts.  Model loading/baking implementations typically do a lot of memory allocation - which could cause stuttering from garbage collection if it happened at render time.  And immutable, prebuilt models and a fast, pre-populated identity hash map avoid complications from concurrent requests for the same model. 
 
I have models with over 2^64 possible states and specialized data structures and handlers to minimize these problems, but I recognize that's not the norm.  Fortunately a simple MixIn for BlockRenderManager will probably be sufficient.  That could be a PR but I'm a little reluctant to encourage dynamic model generation/baking because of the aforementioned pitfalls.

**Comment by asiekierka** [comment]
I'd say "dynamic" model generation/baking belongs outside of API, yes.

**Comment by therealfarfetchd** [comment]
BakedModel doesn't actually have to be baked - it can create a different model based on the passed state in the getQuads method and cache that. Unless they removed that parameter, but I think it's still there.

**Comment by asiekierka** [comment]
It's still there, yes, and it's not going anywhere.

But I've always seen passing "extended" states as a bit of a hack. Maybe we can just get away with passing a fourth parameter + an AdvancedBakedModel itf?

**Comment by grondag** [comment]
Yes, BakedModel can be implemented to generate quads dynamically based on the BlockState parameter passed to getQuads.  

However, because that BlockState is also the key used to find the model in an IdentityHashMap, why bother?  If you have a number of variants that can reasonably fit within the hash map, you may as well pre-bake all of them.  

And if you have too many block states to fit in the hash map, then you will never see those block states as a parameter to getQuads.

Most of the blocks that need this (multi-part, conduit covers, Chisels and Bits) are going to be BlockEntities anyway, so they can (and probably do already) implement a BlockEntityRenderer.

But that is a lost opportunity, performance-wise, because the blocks usually aren't actually dynamic in the world and could be built and rendered as part of the chunk.  Rendering from a pre-packed VBO is always going to be faster than uploading the quads each frame.

**Comment by asiekierka** [comment]
Yes, what I'm saying is: what way should we look into for rendering into a pre-packed VBO, but with arbitrary data?

**Comment by grondag** [comment]
Also, I agree ExtendedState is a hack.  And in 1.13+ it seems better to put as many properties a possible into BlockState itself, with the hope of avoiding the need for BlockEntities.

Assuming the new world format can handle arbitrarily many block states (I haven't worked with it enough yet to know) then, theoretically, we would simply need a way to circumvent the IdentityHashmap in BlockModels and instead route the lookup to a model dispatch mechanism that doesn't rely on a fixed number of block states. 

Putting all the variation into BlockState directly (assuming again that the chunk format can handle it) would also prevent the performance-killing neighbor state or BlockEntity NBT lookups that are otherwise needed to determine the full extended state.

**Comment by asiekierka** [comment]
>it seems better to put as many properties a possible into BlockState itself

NO! The BlockStates themselves *still* take up space in memory - as you need to store transitions from one BlockState to another, which is at *least* a large array + all state combinations possible.

**Comment by grondag** [comment]
Ah, that is good to know.  Disappointing but not surprising.

I suppose we could extend BlockState to handle additional attributes that aren't defined as properties and leave the serialization implications up to mod authors.  But those will obviously not work with the identity-based lookup as-is.  I also don't know if there are other situations where an == comparison needs to work for BlockStates.

Perhaps we do need that fourth parameter with additional attributes.

**Comment by asiekierka** [comment]
@grondag Honestly, the main problem here is that looking up and adding all the information to an "extended state" object is costly as-is, I think we'd rather pass something as direct as possible - but we have to watch out for thread safety...

**Comment by grondag** [comment]
Also agree.  Creating fewer objects is better.  At the same time immutability is attractive for thread safety.

**Comment by asiekierka** [comment]
And we know for a fact that any additional state falls into one of two cases:

* (a) based on neighbouring blocks - which is usually manageable enough to put into a BlockState,
* (b) based on a BlockEntity.

Why not, then, make the "fourth parameter" simply acquired from a BlockEntity instance?

**Comment by grondag** [comment]
I have one case (dynamic terrain) where the number of additional states is on the order of 2^54 and is based on neighboring blocks.  Currently these are implemented without BlockEntities (unless the terrain is "frozen".)

So I'd prefer to have world and position references and then look up BlockEntity if it applies.

**Comment by shadowfacts** [comment]
Adding a `FabricBakedModel` type which has an extra `getQuads` method that accepts a nullable block view + position would solve the vast majority of problems that extended block states are used for in Forge.

All of the usages of extended block states I can think of off the top of my head are used to retrieve data from block entities. This is the approach I take for my multipart mod, a [custom block state](https://github.com/shadowfacts/SimpleMultipart/blob/master/src/main/java/net/shadowfacts/simplemultipart/container/MultipartContainerBlockState.java) class and a [mixin](https://github.com/shadowfacts/SimpleMultipart/blob/master/src/main/java/net/shadowfacts/simplemultipart/mixin/client/MixinBlockRenderManager.java#L38) to pass the "rendering state" into `getQuads` while still using the normal state for `BakedModel` lookup.

Having a block view and position would solve this without the need for a mixin or a custom block state.

But that's an issue separate from model loading.

**Comment by asiekierka** [comment]
Yes, I think it'd be wise to open a separate issue.

**Comment by grondag** [comment]
Getting back to the PR at hand - assuming we do create some sort of extended model dispatch hook, what does it imply for model loading? 

The dynamic generation hook will need a way to "know" if the model requires dynamic handling.  Seems like there are two ways for that to happen:

1) The normal BlockState : BakedModel lookup applies, and then the dynamic hook has to inspect the model to determine if special handling is required. 

2) The dynamic hook has a more direct way of knowing that dynamic model dispatch is needed and routes directly to the appropriate custom dispatcher.  (maybe it could know via Block.blockRenderType? ) 

I suppose there is a third option, mod authors can and should register a model for "loading" so that they can handle parsing, declare textures for the texture map, etc. at the appropriate time for special handling.  But the "load" doesn't have to do anything if it isn't required, and the IdentityHashmap lookup is never used for their models - their custom dispatcher we be called directly.

I think I prefer this third way - I spent a lot of time optimizing the lookup and caching of dynamic models from extended state - I don't like that a useless (albeit fast) HashMap call has to happen before when it always returns the same reference to my dispatcher.

**Comment by asiekierka** [comment]
My assumption was that dynamic models will simply have model loaders which provide UnbakedModels which then create DynamicBakedModels, meaning that any additional hooks are unnecessary. (The way vanilla does things with its blockstate format is likely unsuitable for dynamic models regardless.)

**Comment by asiekierka** [comment]
To add to this point: the HashMap lookup may be useless when it routes to *your* dispatcher, but what if we have *multiple* possible dispatchers?

I suppose we could add an interface to Block which returns a DynamicModelDispatcher and circumvent the whole system that way (not to mention - automatically filter them out from blockstate processing in the model API), but we have to be careful with sidedness here!

**Comment by grondag** [comment]
> To add to this point: the HashMap lookup may be useless when it routes to _your_ dispatcher, but what if we have _multiple_ possible dispatchers?

Well, that would be strange, right?

> I suppose we could add an interface to Block which returns a DynamicModelDispatcher and circumvent the whole system that way (not to mention - automatically filter them out from blockstate processing in the model API), but we have to be careful with sidedness here!

This is more or less what I was imagining.  Though still need a way to declare textures for inclusion in texture map. 

After more thought, it probably is best to keep the first approach, and simply have the model loaders provide unbaked models that implement the necessary interface to trigger dynamic handling.  I say this for three reasons:

1) We don't need a separate mechanism/even for getting textures into the texture map.

2) Reduces special handling elsewhere.  The dynamic hook just has to inspect the type of model it gets.

3) Pack makers could, theoretically, substitute a dynamic model for a static model.

To illustrate #3, suppose I made a dynamic stone or sand model that implements flowing terrain and multi-block textures.  I would need to give the model an "loader" of some kind, but a pack maker could choose to use my dynamic model simply by changing some JSON files.

As much as I dislike the waste, an IdentityHashMap lookup is always going to be fast...

**Comment by grondag** [comment]
@asiekierka I need to step away for a while, just want to say thank you for the excellent work on this PR and the other issue.  Shaping up really well.

**Comment by asiekierka** [comment]
SO, GETTING BACK ON TOPIC

The only real blockers for this PR as-is are "merging accepts() and load()" and "figuring out ModelRequester", right?

In particular, does anyone have any alternate ideas for ModelRequester? Because I still think it's fine as-is.

**Comment by shadowfacts** [comment]
Currently `ModelRequester` will not work with any non-`Block` objects.

When `ModelLoader::loadModel` is called with a given `id`, it first looks for a variant map (from a blockstate definition JSON). But it doesn't just parse each of the variants provided in the state definition into their own `ModelIdentifier`+`UnbakedModel` pairs. Instead, it gets the `Block` with the same `id` and uses that block's state factory to construct a list of states it should load from each given state definition.

The issue appears to stem from `ModelLoader` allowing there to be multiple blockstate definitions, as long as each of them provides disjoint sets of blockstates.

The only solution I've though of so far is to tear out all of `loadModel` and replace it with a Fabric implementation that's independent of block/blockstates and instead just generates `ModelIdentifier` -> `UnbakedModel` entries for every variant that's present in the blockstate definitions. This might require removing the ability for blockstate definitions to be split across multiple resource packs (which, as it's removing a vanilla feature, would be sub-optimal).

This solution is admittedly very invasive, so I'd welcome suggestions for any other potential fixes for this problem.

**Comment by asiekierka** [comment]
> Edit: If supplier/function are for saving memory, a setup event for (re)triggering the registrations and throwing them away after finishing is probably better. I don't think it'd save much though, assuming reasonable implementations.

It's to allow loaders to cache during the loading process. Every time ModelLoader is re-instantiated, you can pretty much throw all assumptions about the resource pack contents out the window - making it a supplier makes it easy to implement caching without worrying about implementing cache invalidation.

> MC has a global model id -> baked model repo that can be directly queried at runtime to build those caches. Post bake events can be used likewise.

The idea here is that each ModelRequester only accesses its own area of interest; so that modders don't "cheat", make a dummy ModelRequester and then read all of the baked models out. However, it *is* true that all of the baked models are stored regardless... It can be dropped, then, and replaced with a post-model-loading event.

**Comment by asiekierka** [comment]
I have pushed some tweaks to the API. Of note:

* ModelProvider is the new "loader"; it also has a way to load parent models (vanilla *does* gracefully handle circular references),
* ModelAppender is a much simplified take on the old ModelRequester,
* In production (so no development and no debug flag enabled), only the first UnbakedModel is used,
* An optimized loading mechanism is used with either approach for custom loaders (though the JsonModelProvider optimization is *not* present yet).

**Comment by asiekierka** [comment]
From chat:

```shadowfactsToday at 3:44 PM
tl;dr: it's not usable for non-Block objects
asieToday at 3:44 PM
But... it is?
> When ModelLoader::loadModel is called with a given id, it first looks for a variant map (from a blockstate definition JSON).
No. It first calls the ModelProviders.
And ModelProviders now have a way to call loadModel back, with a different Identifier.
So you can circumvent the whole logic.
Though it should be documented that ModelIdentifiers are for resolving models and Identifiers are for loading model files...
Or even the ModelProvider split into two based on this dichotomy. Feel free to write up some thoughts on that if you have any.```

**Comment by asiekierka** [comment]
Rough proposal:

* ModelResolvers take in a ModelIdentifier and return an Identifier or UnbakedModel.
* ModelProviders take in an Identifier and return an UnbakedModel.

? (Not sure if the split is needed or if we shouldn't just document it. You can already implement all of this as ModelProviders can now call loadModel recursively - Mojang takes care of detecting circular references)

File: src/main/java/net/fabricmc/fabric/api/client/model/ModelProvider.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.client.model;
+
+import com.sun.istack.internal.Nullable;
```
Comment (shadowfacts) [inline src/main/java/net/fabricmc/fabric/api/client/model/ModelProvider.java]: Shouldn't be using this `Nullable`, should be the `javax.annotation` one (which, for some reason, seems not to be present).

Comment (shadowfacts) [reply src/main/java/net/fabricmc/fabric/api/client/model/ModelProvider.java]: This was the one that was suggested: https://github.com/apiguardian-team/apiguardian

Looks fine to me.

**Comment by grondag** [comment]
I like the simplification of ModelAppender.  Use cases are making more sense to me now.

File: src/main/java/net/fabricmc/fabric/mixin/client/model/MixinModelLoader.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+	}
+
+	@Shadow
+	private void putModel(Identifier id, UnbakedModel unbakedModel) {
+
+	}
+
+	@Inject(at = @At("HEAD"), method = "loadCustomModel", cancellable = true)
+	private void loadModel(Identifier id, CallbackInfo ci) {
```
Comment (shadowfacts) [inline src/main/java/net/fabricmc/fabric/mixin/client/model/MixinModelLoader.java]: You have the method name and the mixin injection name switched here.

Comment (asiekierka) [reply src/main/java/net/fabricmc/fabric/mixin/client/model/MixinModelLoader.java]: nice refactor right there

File: src/main/java/net/fabricmc/fabric/api/client/model/ModelProvider.java [hunk 3]
```diff
@@ -49,5 +48,5 @@
 	 * @return The loaded UnbakedModel, or null if this ModelProvider doesn't handle a specific Identifier
 	 * (or if there was no error!).
 	 */
-	@Nullable UnbakedModel load(Identifier id, Context context) throws ModelProviderException;
+	/* @Nullable */ UnbakedModel load(Identifier id, Context context) throws ModelProviderException;
```
Comment (shadowfacts) [inline src/main/java/net/fabricmc/fabric/api/client/model/ModelProvider.java]: There's another occurrence of this issue in ModelLoadingRegistryImpl

File: src/main/java/net/fabricmc/fabric/api/client/model/ModelResourceProvider.java [hunk 4]
```diff
@@ -37,8 +37,8 @@
  * easily conflict with another OBJ loader unless you take some precautions,
  * for example:
  *
- * a) Only load files with a mod-suffixed name, such as .architect.obj,
- * b) Only load files from an explicit list of namespaces, registered elsewhere.
+ * a) Only loadModelResource files with a mod-suffixed name, such as .architect.obj,
+ * b) Only loadModelResource files from an explicit list of namespaces, registered elsewhere.
```
Comment (shadowfacts) [inline src/main/java/net/fabricmc/fabric/api/client/model/ModelResourceProvider.java]: I don't think that's what you intended to do :P

File: src/main/java/net/fabricmc/fabric/mixin/client/texture/MixinSpriteAtlasTexture.java [hunk 5]
```diff
@@ -149,7 +149,7 @@ public void loadSprite(ResourceManager manager, Sprite sprite, CallbackInfoRetur
 					return;
 				}
 			} catch (RuntimeException | IOException e) {
-				LOGGER.error("Unable to load custom sprite {}: {}", sprite.getId(), e);
+				LOGGER.error("Unable to loadModelResource custom sprite {}: {}", sprite.getId(), e);
```
Comment (shadowfacts) [inline src/main/java/net/fabricmc/fabric/mixin/client/texture/MixinSpriteAtlasTexture.java]: Or here

Comment (asiekierka) [reply src/main/java/net/fabricmc/fabric/mixin/client/texture/MixinSpriteAtlasTexture.java]: ugh intellij

Comment (liach) [reply src/main/java/net/fabricmc/fabric/mixin/client/texture/MixinSpriteAtlasTexture.java]: yeah, try using refactor method instead.

File: src/main/java/net/fabricmc/fabric/api/client/model/ModelAppender.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+package net.fabricmc.fabric.api.client.model;
+
+import net.minecraft.client.util.ModelIdentifier;
+import net.minecraft.resource.ResourceManager;
+
+import java.util.function.Consumer;
+
+@FunctionalInterface
+public interface ModelAppender {
+	void append(ResourceManager manager, Consumer<ModelIdentifier> modelAdder);
```
Comment (sfPlayer1) [inline src/main/java/net/fabricmc/fabric/api/client/model/ModelAppender.java]: -> void appendAll(ResourceManager manager, Consumer<ModelIdentifier> out); for clarity

Comment (asiekierka) [reply src/main/java/net/fabricmc/fabric/api/client/model/ModelAppender.java]: I'd say "appendAll" is a bit confusing, to me, and would rather go with "appendModels" or "appendAllModels" or "appendIds" or "appendAllIds" but I'll ask for a second opinion
